### ISSUE TITLE
Account restricted from Redeeming Rewards

### ISSUE THREAD

**Issue by shopbuy11** [charlesbel/Microsoft-Rewards-Farmer#242]
This is not a bug of the bot per se, but apparently MSFT has implemented something that makes the bot less useful now...

Until recently (back in June), I had used 5 accounts for years and farmed and redeemed quite a bit of reward points.    Then all the sudden all 5 accounts got banned completely.   So of course I just created a few new accounts. :)  Initially they worked fine - I could farm and redeem points.   Then starting a couple of weeks ago, whenever I tried to redeem, it just says that account is restricted from redeeming.   They all can farm and earn points daily without issue, but they can't redeem.  Does this happen to anyone else?   Is it happening newly created accounts only?   Any way around this?  TIA.

![image](https://github.com/charlesbel/Microsoft-Rewards-Farmer/assets/136040705/c075f5d4-b8d0-48d8-8da9-8a6ac3f668aa)

**Comment by tony11306** [comment]
it looks like it's not newly created account only, cuz my main account, which did run the script, got restricted as well

**Comment by udu3324** [comment]
I got banned, started using it legit for 2 ish days, unbanned

**Comment by joshuamooree** [comment]
I started getting the same issue.  Not sure exactly what the cause is, but selenium is detectable by websites, so that could be at least a part of it.  There's a modified version of chromedriver that should be more stealthy.  I created a fork to use that and created a pull request: #248

**Comment by shopbuy11** [comment]
> I started getting the same issue. Not sure exactly what the cause is, but selenium is detectable by websites, so that could be at least a part of it. There's a modified version of chromedriver that should be more stealthy. I created a fork to use that and created a pull request: #248 . Do note though that this adds a new dependency. I allow the script to work without undetected_chromedriver, but you obviously don't get the stealth if you do that.

Thanks a lot.  I replaced the main.py with your branch.  Is that enough?   Do I need to use a different version of ChromeDriver?  If so, where can I get it.

**Comment by shopbuy11** [comment]
> I got banned, started getting points without the bot for 2 ish days, unbanned

Will try that and see if it works for me.  Thanks!

**Comment by joshuamooree** [comment]
> > I started getting the same issue. Not sure exactly what the cause is, but selenium is detectable by websites, so that could be at least a part of it. There's a modified version of chromedriver that should be more stealthy. I created a fork to use that and created a pull request: #248 . Do note though that this adds a new dependency. I allow the script to work without undetected_chromedriver, but you obviously don't get the stealth if you do that.
> 
> Thanks a lot. I replaced the main.py with your branch. Is that enough? Do I need to use a different version of ChromeDriver? If so, where can I get it.

No, the files you really need are src/browser.py and src/userAgentGenerator.py.  src/browser.py is where the selenium instance gets created, so that's the heart of the change.  But undetected-webdriver is a bit more fussy with how you set user agent than normal webdriver, so I had to add some stuff to userAgentGenerator.py for that.  The new version is also a bit more stealthy because there's some headers that Chrome sends that have additional identifying info in addition to the user agent.  The new version should change those headers to be closer to what Edge uses.  I'd recommend just replacing all the .py files since there's changes to use logging instead of print statements in several of them (should still work, you just won't get some of the progress stuff logged to a text file).  To get the alternate chromedriver, just run "pip install -r requirements.txt" again (with my modified requirements.txt, of course) and it'll get pulled in automatically.  If it's all working correctly, you'll see an extra line at startup noting that the chromedriver is being patched (something undetected-chromedriver does, not me).

**Comment by shopbuy11** [comment]
Thanks again.  I shall try and report results.

**Comment by charlesbel** [comment]
Just to let everybody know I am currently writing over @joshuamooree PR to enhance even better what he done. Will soon publish it. Thanks for everybody's patience.

**Comment by joshuamooree** [comment]
> Just to let everybody know I am currently writing over @joshuamooree PR to enhance even better what he done. Will soon publish it. Thanks for everybody's patience.

One thing I would do (have been meaning to do on my fork) is check the SEC-CH-UA-* headers between desktop Edge and Mobile Edge.  I'm not 100% convinced that I have it emulating them right.  Pretty sure SEC-CH-UA is right (note that the version number for Not.A/Brand is technically supposed to be random along with the order for the items listed in that header, but I have it hard coded), but there's a bunch of other ones (SEC-CH-UA-ARCH, SEC-CH-UA-MOBILE, SEC-CH-UA-MODEL, and others) that need checking.

**Comment by tony11306** [comment]
Here to report my results. For these two or three days, I've tested account that doesn't run with the script, and like what @udu3324 said, my main account is able to redeem now. I also tested my sub account that runs the undetected version script, and it also let me redeem rewards now!(only tested on the "save the children" donation option)

**Comment by charlesbel** [comment]
Updated the script ! Please try it and give feedbacks, I'll let this issue open so we can track how things are going. Thanks @joshuamooree !

**Comment by udu3324** [comment]
The bot crashes if It doesn't run with -v, and it spits out the error below

[ERROR] TypeError: '<' not supported between instances of 'NoneType' and 'int'

**Comment by charlesbel** [comment]
> The bot crashes if It doesn't run with -v, and it spits out the error below
> 
> [ERROR] TypeError: '<' not supported between instances of 'NoneType' and 'int'

Opened an issue for that
